Arale modem FW switch

alfonsosanchezbeato opened the pull request:
Load modem FW depending on MCC/MNC from the SIM. This is needed because different modem FWs support different radio technologies (due possibly to modem memory restrictions), and depending on the operator's network one or another technology is required. Additionally, we use radio power in case we have a single-slot modem.

alfonsosanchezbeato commented:
* Tested on arale (vivid 46) and on krillin (rtm promoted 13). In the case of krillin, QUERY_MODEM_TYPE returns 3 and no attempt to reload fw is performed.

* To test this on arale: change file /data/nvram/APCFG/APRDCL/MD_Type contents to the following 6 bytes:

00 00 00 00 aa 00

make sure that owner/permissions are the same as the original. Then reboot. ofonod should detect a wrong FW, change it and force a modem reset. After that, contents of MD_Type should be:

05 00 00 00 aa 05

vicamo commented on plugins/mtk.c:
```diff
@@ -326,6 +334,142 @@ static void sim_inserted(struct ril_msg *message, gpointer user_data)
+	g_mtk_request_set_trm(md->ril, 0x0B, &rilp);
+
+	if (g_ril_send(md->ril, MTK_RIL_REQUEST_SET_TRM, &rilp,
+			set_trm_cb, modem, NULL) == 0)
```
From my Android log dump, there is no MTK_RIL_REQUEST_SET_TRM ever been sent, but there are STORE_MODEM_TYPE ones. It follows that after STORE_MODEM_TYPE responses, it's not necessarily followed by a MTK_RIL_REQUEST_SET_TRM req. Sending this request has no effect on my tests, but the two lines
```C
md->ofono_online = FALSE;
ofono_modem_set_powered(md->modem, FALSE);
```
in the `set_trm_cb` will cause the device cease to register to the network.

alfonsosanchezbeato commented:
@vicamo MTK_RIL_REQUEST_SET_TRM is sent when a FW switch is done. If you want to see it in Android, reflash the phone with Android (better if you do it with the engineering build so you can modify easily nvram files), modify MD_Type with values "00 00 00 00 aa 00" as explained in previous comments and reboot the phone (do not forget to get radio log with "wait-for-device" so you get all radio log). You will see that SET_TRM is issued. If that is not the case your phone exhibits a different behaviour from mine, which is quite worrying.

Second, SET_TRM is meant to be sent only when a FW switch is done and resets the modem. It should not be sent otherwise. And the lines you mention are basically saying ofono that the modem is off. When I have tested this, SET_TRM resets the modem and rild closes the socket. This event is detected by mtk.c, and we try to reconnect a few seconds later.

vicamo commented:
With an updated deb from Alfonso, this pull request, along with others which are only available in bzr branch, was verified with CHT 4G SIM card in Taiwan.

alfonsosanchezbeato commented:
Re-based

tonyespy commented:
A couple of quick comments...

1. The first commit "add RIL_REQUEST_STORE_MODEM_TYPE" is lacking a commit message body.  This is probably OK as I see you then re-worked the code in your subsequent commits.

2. The "Add parcels..." commit message isn't 100% accurate as the store parcel code was actually added the first commit.

3. Didn't we recently get access to the MTK AT modem documentation?  Is so, is it possible for us to determine what the magic #8 represetns in the TRM request?

4. No unit test for STORE_MODEM_TYPE.

5. set_trm_cb() has a new TODO which states that we should probably adjust the reconnect timer for arale.  Why not do this now?

6. Other than that, the code looks good.  One thing we might consider however is moving the mapping tables into a text file vs. hard-coding them.  I realize that new operators coming to market probably isn't all that common in China, but the current design would require a new ofono release in order to support a new operator.

alfonsosanchezbeato commented:
Re-pulled addressing comments.

1,2.- Commits comments amended. Also moved all changes to mtk.c to only one commit.
3.- Unfortunately I have found no information in the manual about SET_TRM. In fact, the request does not actually trigger any AT command. My guess is that SET_TRM maybe sets some IO pin that resets the modem.
4.- There is actually one test for STORE_MODEM_TYPE, check test-mtkrequest.c
5.- I have measured the time that takes rild to re-start after issuing SET_TRM. It is around 5 seconds. Now we have 8 seconds before trying to reconnect. Although there are 3 seconds of difference, it is probably better to keep those seconds as safeguard, as making it tighter would not make such a difference and times could vary. So I have removed the comment.
6.- Agreed, it might be interesting to make that more configurable. Note however that the codes were also hard-coded in the Android code.

tonyespy commented:
Approved.
